src/irc_bridge.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the IRC bridge class to represent the matrix-appservice-app plugin for Synapse."""
+
+# disabling due the fact that collect status does many checks
+# pylint: disable=too-many-return-statements
+
+import logging
+
+import ops
+from ops.pebble import APIError, ExecError
+
+import pebble
+import synapse
+from charm_state import CharmBaseWithState, CharmState, inject_charm_state
+
+logger = logging.getLogger(__name__)
+
+IRC_SERVICE_NAME = "irc"
+
+
+class PEMCreateError(Exception):
+    """An exception raised when the PEM file creation fails."""
+
+    def __init__(self, message: str):
+        """Initialize a new instance of the PEMCreateError class.
+
+        Args:
+            message: The error message.
+        """
+        super().__init__(message)
+
+
+class IRCBridgeObserver(ops.Object):  # pylint: disable=too-few-public-methods
+    """A class representing the IRC bridge plugin for Synapse application.
+
+    See https://github.com/matrix-org/matrix-appservice-irc/ for more details about it.
+    """
+
+    def __init__(self, charm: CharmBaseWithState) -> None:
+        """Initialize a new instance of the IRC bridge class.
+
+        Args:
+            charm: The charm object that the IRC bridge instance belongs to.
+        """
+        super().__init__(charm, "irc-bridge")
+        self._charm = charm
+        self.framework.observe(charm.on.collect_unit_status, self._on_collect_status)
+
+    def get_charm(self) -> CharmBaseWithState:
+        # pylint:disable=duplicate-code
+        # pylint is extra picky over the smallest details
+        # this doesn't need refactoring
+        """Return the current charm.
+
+        Returns:
+            The charm object.
+        """
+        return self._charm
+
+    @inject_charm_state
+    def _on_collect_status(self, event: ops.CollectStatusEvent, charm_state: CharmState) -> None:
+        # pylint:disable=duplicate-code
+        # the code that is duplicated checks that Synapse is ready
+        # the relevant lines are three and refactoring this in a separate method
+        # would not add any value
+        """Collect status event handler.
+
+        Args:
+            event: Collect status event.
+            charm_state: Instance of CharmState.
+        """
+        if not charm_state.synapse_config.enable_irc_bridge:
+            return
+        if charm_state.irc_bridge_datasource is None:
+            ops.MaintenanceStatus("Waiting for irc bridge db relation")
```

Comment:
i think you meant to put `self.unit.status` = :D

Reply:
@yanksyoon you have the eye of the :tiger: